newrelic/api/opentelemetry.py

```diff
@@ -26,29 +30,77 @@
 from newrelic.api.message_trace import MessageTrace
 from newrelic.api.message_transaction import MessageTransaction
 from newrelic.api.time_trace import current_trace, notice_error
-from newrelic.api.transaction import Sentinel, current_transaction
+from newrelic.api.transaction import Sentinel, current_transaction, accept_distributed_trace_headers, insert_distributed_trace_headers
 from newrelic.api.web_transaction import WebTransaction
+
 from newrelic.core.otlp_utils import create_resource
 
 _logger = logging.getLogger(__name__)
 
 
+class NRTraceContextPropagator(TraceContextTextMapPropagator):
+    HEADER_KEYS = ("traceparent", "tracestate", "newrelic")
+
+    def extract(self, carrier, context=None, getter=None):
+        transaction = current_transaction()
+        # If we are passing into New Relic, traceparent 
+        # and/or tracestate's keys also need to be NR compatible.
+        
+        if transaction:
+            nr_headers = {header_key: ("" if not getter.get(carrier, header_key) else getter.get(carrier, header_key)[0]) for header_key in self.HEADER_KEYS}
+            transaction.accept_distributed_trace_headers(nr_headers)
+        return super().extract(carrier=carrier, context=context, getter=getter)
+    
+
+    def inject(self, carrier, context=None, setter=None):
+        transaction = current_transaction()
+        # Only insert headers if we have not done so already this transaction
+        # New Relic's Distributed Trace State will have the following states:
+        #   0 (00) if not set:
+        #       Transaction has not inserted any outbound headers nor has
+        #       it accepted any inbound headers (yet).
+        #   1 (01) if already accepted:
+        #       Transaction has accepted inbound headers and is able to 
+        #       insert outbound headers to the next app if needed.
+        #   2 (10) if inserted but not accepted:
+        #       Transaction has inserted outbound headers already.
+        #       Do not insert outbound headers multiple times. This is
+        #       a fundamental difference in OTel vs NR behavior: if
+        #       headers are inserted by OTel multiple times, it will
+        #       propagate the last set of data that was inserted. NR
+        #       will not allow more than one header insertion per
+        #       transaction.
+        #   3 (11) if accepted, then inserted:
+        #       Transaction has accepted inbound headers and has inserted
+        #       outbound headers.
+
+        if not transaction:
+            return super().inject(carrier=carrier, context=context, setter=setter)
+        
+        if transaction._distributed_trace_state < 2:
+            nr_headers = []
+            transaction.insert_distributed_trace_headers(nr_headers)
+            for key, value in nr_headers:
+                setter.set(carrier, key, value)
```

Comment:
This looks good now-so much simpler!

A couple additional thoughts/considerations here:

What if the carrier already has that header in it? I think it makes sense to override the parent header but it might make more sense to add to/combine the tracestate header value? I'm not sure how we would add to it though given that we don't have a getter function so this is prob fine.

[This](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/trace/propagation/tracecontext.py#L104) overrides the parent header value that we generate - I assume this is what is called when we super - is the override always guaranteed to match the value we generate?

Reply:
I'm having a flashback to why I had omitted the `super()` call in for this specific condition.  It will not hurt anything (the results for traceparent and tracestate are the same or that the results for traceparent are the same and that the OTel span does not have a tracestate whereas NR passed in its own tracestate, so that result does not get overridden).  It's just redundant.

I'll put in a comment here so that I don't forget why I excluded it (again).